File: recipes/context/typescript/ci.ts
import "dotenv/config";
import assert from "node:assert/strict";
import { context, entityId } from "@persql/context";
import { PerSQL, type PerSQLDatabase } from "@persql/sdk";

const token = process.env.PERSQL_TOKEN;
const database = process.env.PERSQL_DATABASE;
const runId = process.env.GITHUB_RUN_ID ?? `local-${Date.now()}`;

const mode = token ? "remote" : "local";
console.log(`[ci] mode=${mode}`);

// Remote: claim a fresh branch per run for isolation, delete it on exit.
// Local:  in-process :memory: SQLite, no network, no cost.
let handle: PerSQLDatabase;
let cleanup: (() => Promise<void>) | undefined;

if (token && database) {
  const parent = new PerSQL({ token });
  const branchRef = process.env.PERSQL_BRANCH_REF ?? `ci-${runId}`;
  const claimed = await parent.database(database).branches.claim({
    ref: branchRef,
    role: "admin",
    ttlSec: 3600,
  });
  handle = new PerSQL({ token: claimed.token }).database(
    `${claimed.namespaceSlug}/${claimed.databaseSlug}`
  );
  cleanup = async () => {
    await parent.database(database).branches.delete(branchRef);
    console.log("[ci] branch cleaned up");
  };
} else {
  handle = new PerSQL({ local: ":memory:" }).database("test/ci");
}

try {
  const ctx = context(handle, { source: "ci-agent" });
  await ctx.init();

  // 1. Remember three memories, confirm they land in the always-loaded index.
  await ctx.remember({
    name: "billing-model",
    description: "How customers pay",
    type: "project",
    body: "Usage is metered into a prepaid balance, topped up via Checkout.",
  });
  await ctx.remember({
    name: "reply-tone",
    description: "Voice for replies",
    type: "user",
    body: "Short, declarative, no exclamation marks.",
  });
  await ctx.remember({
    name: "orders-schema",
    description: "Orders table shape",
    type: "reference",
    body: "orders(id, customer_id, total_cents, status).",
  });
  const index = await ctx.index();
  assert.equal(index.length, 3, `expected 3 memories, got ${index.length}`);
  assert.deepEqual(
    [...index.map((m) => m.name)].sort(),
    ["billing-model", "orders-schema", "reply-tone"],
    `unexpected index: ${index.map((m) => m.name)}`
  );
  console.log("[ci] remember + index ok");

  // 2. Lexical recall (FTS5/BM25) surfaces the most relevant memory first.
  const hits = await ctx.recall("prepaid balance checkout");
  assert.ok(hits.length >= 1, "recall returned nothing");
  assert.equal(hits[0].name, "billing-model", `wrong top hit: ${hits[0]?.name}`);
  console.log("[ci] recall ok");

  // 3. Same name UPSERTs — the body changes, the row count does not.
  await ctx.remember({
    name: "billing-model",
    description: "How customers pay",
    type: "project",
    body: "Prepaid balance with a one-time welcome credit on first use.",
  });
  const afterUpsert = await ctx.index();
  assert.equal(afterUpsert.length, 3, "upsert created a duplicate row");
  const updated = await ctx.get("billing-model");
  assert.ok(updated?.body.includes("welcome credit"), "upsert did not change body");
  console.log("[ci] upsert by name ok");

  // 4. Entity graph: link two entities, walk the neighbourhood. Edges store
  //    entity ids (entityId(name)), not the raw names.
  await ctx.link("billing-meter", "writes_to", "balance-ledger");
  const { edges } = await ctx.neighbors("billing-meter", { depth: 1 });
  assert.ok(
    edges.some(
      (e) =>
        e.src === entityId("billing-meter") &&
        e.dst === entityId("balance-ledger") &&
        e.rel === "writes_to"
    ),
    "linked edge not found in neighbourhood"
  );
  console.log("[ci] link + neighbors ok");

  // 5. Forget removes the memory by name.
  await ctx.forget("orders-schema");
  assert.equal(await ctx.get("orders-schema"), null, "forget did not delete");

  // 6. A fresh store reads the persisted state — memory lives in the database,
  //    not the wrapper. Two memories survive, one edge, two entities.
  const fresh = context(handle);
  const stats = await fresh.stats();
  assert.equal(stats.memories, 2, `expected 2 memories, got ${stats.memories}`);
  assert.equal(stats.edges, 1, `expected 1 edge, got ${stats.edges}`);
  assert.equal(stats.entities, 2, `expected 2 entities, got ${stats.entities}`);
  console.log("[ci] forget + reread by fresh store ok");

  console.log(`[ci] PASS (${mode})`);
} finally {
  await cleanup?.();
}

File: recipes/hermes-memory/python/ci.py
"""Headless test: Hermes agent memory persists via PerSQL.

Local mode  (default, no secrets): PerSQL(local=":memory:") — proves the
  memory store contract with no network, no LLM.
Remote mode (PERSQL_TOKEN + PERSQL_DATABASE set): claims a fresh branch per
  run for isolation, runs the same checks, deletes the branch on exit.

`hermes-persql` is a Hermes memory provider; under the hood it reads and writes
a PerSQLMemoryStore. Hermes calls sync_turn -> record_turn, prefetch /
persql_recall -> recall, and persql_remember -> remember. This script exercises
that store the way the provider does.

Exit 0 on pass, non-zero on failure. No interactive input.
"""

from __future__ import annotations

import os
import time

from dotenv import load_dotenv
from persql import PerSQL

from hermes_persql import PerSQLMemoryStore

load_dotenv()

TOKEN = os.environ.get("PERSQL_TOKEN")
DATABASE = os.environ.get("PERSQL_DATABASE")
RUN_ID = os.environ.get("GITHUB_RUN_ID") or str(int(time.time()))

mode = "remote" if TOKEN else "local"
print(f"[ci] mode={mode}")


def main() -> None:
    cleanup = None

    if TOKEN and DATABASE:
        parent = PerSQL(token=TOKEN)
        branch_ref = os.environ.get("PERSQL_BRANCH_REF") or f"ci-{RUN_ID}"
        claimed = parent.database(DATABASE).branches.claim(
            ref=branch_ref, role="admin", ttl_sec=3600
        )
        db = PerSQL(token=claimed["token"]).database(
            f"{claimed['namespace_slug']}/{claimed['database_slug']}"
        )

        def cleanup() -> None:
            parent.database(DATABASE).branches.delete(branch_ref)
            print("[ci] branch cleaned up")
    else:
        db = PerSQL(local=":memory:").database("test/ci")

    try:
        store = PerSQLMemoryStore(db)
        store.ensure_setup()

        # 1. A turn (Hermes sync_turn) is persisted and recalled by keyword.
        store.record_turn("s1", "What is PerSQL?", "An isolated SQLite database per agent, on the edge.")
        assert any("edge" in h for h in store.recall("PerSQL edge")), "turn not recalled"
        print("[ci] turn recorded and recalled (FTS5)")

        # 2. A durable fact (Hermes persql_remember) is stored, listed, recalled.
        store.remember("The user prefers PerSQL for agent memory.", session_id="s1")
        assert store.list_facts() == ["The user prefers PerSQL for agent memory."], "fact not listed"
        assert any("PerSQL" in h for h in store.recall("agent memory")), "fact not recalled"
        print("[ci] fact stored, listed, and recalled")

        # 3. A fresh store over the same database replays everything — survives a
        #    process restart, not an in-memory cache.
        resumed = PerSQLMemoryStore(db)
        assert resumed.recent_turns("s1"), "memory not found after restart"
        print("[ci] memory recovered by a fresh store")

        # 4. Recall is lexical and bounded — an unrelated query returns nothing.
        assert resumed.recall("quantum chromodynamics") == [], "recall not bounded"
        print("[ci] recall bounded")

        print(f"[ci] PASS ({mode})")
    finally:
        if cleanup:
            cleanup()


if __name__ == "__main__":
    main()
